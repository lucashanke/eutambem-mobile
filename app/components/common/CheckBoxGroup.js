import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { View, Text, StyleSheet } from 'react-native';

import { DARK_PURPLE } from '../../styles';
import CheckBox from './CheckBox';
import input from '../hoc/input';

const styles = StyleSheet.create({
  container: {
    marginTop: 10,
    marginBottom: 10,
  },
  text: {
    fontSize: 14,
    color: DARK_PURPLE,
  },
  checkbox: {
    paddingTop: 10,
  },
});

class CheckBoxGroup extends Component {
  constructor(props) {
    super(props);
  }

  onClick = (key) => {
    let selectedItems = this.props.values;
    if (this.props.values.includes(key)) {
      selectedItems = selectedItems.filter(item => item !== key);
    } else {
      selectedItems = selectedItems.concat(key);
    }
    this.props.onValueChange(selectedItems);
  }

  renderOptions = () => {
    return this.props.options.map(option => (
      <CheckBox
        key={option.value}
        onValueChange={() => this.onClick(option.value)}
        value={this.props.values.includes(option.value)}
        label={option.label}
      />
    ));
  }

  render() {
    return (
      <View style={styles.container}>
        <Text style={styles.text}>{this.props.label}</Text>
        {this.renderOptions()}
      </View>
    );
  }
}

CheckBoxGroup.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    key: PropTypes.string,
  })).isRequired,
  label: PropTypes.string.isRequired,
  onValueChange: PropTypes.func.isRequired,
};

export default input(CheckBoxGroup);
