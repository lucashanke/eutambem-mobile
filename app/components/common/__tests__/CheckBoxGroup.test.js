import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';

import { CheckBoxGroup } from '../CheckBoxGroup';

const props = {
  label: 'Test CheckBoxGroup',
  options: [
    {
      value: '1',
      label: 'Label 1',
    },
    {
      value: '2',
      label: 'Label 2',
    },
  ],
  values: ['2'],
  onValueChange: () => {},
};

describe('CheckBoxGroup', () => {
  it('renders a checkbox component for each option passed', () => {
    const wrapper = shallow(<CheckBoxGroup {...props} />);

    expect(wrapper.find('CheckBox')).toHaveLength(2);
  });

  it('renders a checkbox as marked if values prop contains its value', () => {
    const wrapper = shallow(<CheckBoxGroup {...props} />);
    const markedCheckBox = wrapper.find({ value: true });

    expect(markedCheckBox).toHaveLength(1);
    expect(markedCheckBox.props().label).toEqual('Label 2');
  });

  it('when an unmarked option changes, onValueChange prop is called with the selected values', () => {
    const onValueChange = sinon.spy();
    const wrapper = shallow(<CheckBoxGroup
      {...props}
      onValueChange={onValueChange}
    />);

    const unmarkedCheckBox = wrapper.find({ value: false, label: 'Label 1' });
    unmarkedCheckBox.props().onValueChange();

    expect(onValueChange.callCount).toEqual(1);
    expect(onValueChange.calledWith(['2', '1'])).toBeTruthy();
  });

  it('when an unmarked option changes, onValueChange prop is called with the selected values', () => {
    const onValueChange = sinon.spy();
    const wrapper = shallow(<CheckBoxGroup
      {...props}
      onValueChange={onValueChange}
    />);

    const unmarkedCheckBox = wrapper.find({ value: true, label: 'Label 2' });
    unmarkedCheckBox.props().onValueChange();

    expect(onValueChange.callCount).toEqual(1);
    expect(onValueChange.calledWith([])).toBeTruthy();
  });
});
