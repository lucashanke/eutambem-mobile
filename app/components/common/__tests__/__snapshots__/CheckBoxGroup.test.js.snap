// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBoxGroup renders correctly 1`] = `
<ForwardRef(View)
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <ForwardRef(View)
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <CheckBoxGroup
      label="Test CheckBoxGroup"
      onValueChange={[Function]}
      options={
        Array [
          Object {
            "key": "1",
            "label": "Label 1",
          },
          Object {
            "key": "2",
            "label": "Label 2",
          },
        ]
      }
      required={false}
    />
  </ForwardRef(View)>
</ForwardRef(View)>
`;
